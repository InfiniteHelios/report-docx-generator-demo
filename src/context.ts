"Thu Nov 19 2015 11:37:00 GMT-0800 (Pacific Standard Time)" 
