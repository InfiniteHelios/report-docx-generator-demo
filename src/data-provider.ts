"Mon Oct 12 2015 21:35:00 GMT-0700 (Pacific Daylight Time)" 
