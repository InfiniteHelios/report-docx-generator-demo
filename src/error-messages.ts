export const InputDataNotProvided = "Input data was not provided";
export const TemapltePathIsRequired = "Template path must be provided, under general section";