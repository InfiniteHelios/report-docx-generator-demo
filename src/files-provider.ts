"Sat Aug 08 2015 10:34:00 GMT-0700 (Pacific Daylight Time)" 
