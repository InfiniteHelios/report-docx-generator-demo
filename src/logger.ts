"Fri Aug 07 2015 11:56:00 GMT-0700 (Pacific Daylight Time)" 
