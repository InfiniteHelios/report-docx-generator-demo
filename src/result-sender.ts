import Logger from "./logger";
import axios from 'axios';

export class ResultSender {
  private accountId: any;
  private reportId: any;
  private _logger;

  constructor(accountId: any, reportId: any) {
    this.accountId = accountId;
    this.reportId = reportId;
    this._logger = new Logger("ResultSender");
  }

  updatePath(docPath: string) {
    this._logger.debug("Updating report status and path", { path: docPath });
    return this.put(this.accountId, `api/report/update-path`, {
      reportId: this.reportId,
      docPath: docPath
    });
  }

  updateStatusFailed() {
    return this.put(this.accountId, `api/report/update-status-faild`, { reportId: this.reportId });
  }

  updateStatusStarted() {
    return this.put(this.accountId, `api/report/update-status-started`, {
      reportId: this.reportId
    });
  }

  private put(accountId: any, path: any, data: any) {
    this._logger.debug("Going to send put request", { path, data });
    const strData = JSON.stringify(data);
    return new Promise((resolve, reject) => {
      axios
        .put(`${process.env.HOST_URL}/${path}`, 
        strData,
        {
          headers: {
            "Content-Length": strData.length,
            "Content-Type": "application/json",
            secret: process.env.HOST_SECRET,
            account: accountId
          }
        })
        .then((res) => {
          console.log(`statusCode: ${res.statusText}`);
          resolve(res.data);
        })
        .catch((error) => {
          console.error(error);
          reject(false);
        });
    });
  }

}
