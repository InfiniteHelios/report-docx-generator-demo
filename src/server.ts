"Thu Dec 17 2015 22:42:00 GMT-0800 (Pacific Standard Time)" 
