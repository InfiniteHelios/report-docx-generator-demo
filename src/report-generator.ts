import { Placeholders } from "./data/Placeholders";
import { ProjectFields } from "./data/ProjectFields";
import { ItemsStruct } from "./data/ItemsStruct";
import { ReadyComponents } from "./data/ReadyComponents";
import { Items } from "./data/Items";
import { StyleFields } from "./data/Style";
import path from "path";
import { FilesProvider } from "./files-provider";
import { TemplateDownloader } from "./templates-downloader";
import { dirSync } from "tmp";
import { Context } from "./context";
import { DataProvider } from "./data-provider";
import { ResultSender } from "./result-sender";
import fs from "fs";
import { S3 } from "aws-sdk";
import Logger from "./logger";
import { Formulas } from "./data/Formulas";
import { InputDataNotProvided, TemapltePathIsRequired } from "./error-messages";

const INPUT_FOLDER = "simple-table";

export interface IReportResult {
  valid: boolean;
  reportId?: string;
  filename?: string;
  content?: String;
  errorMessage?: string;
}

export class ReportGenerator {
  private dataProvider: DataProvider;
  private filesProvider: FilesProvider;
  private resultSender: ResultSender;
  private _context: Context;
  private _logger;
  private testData;

  constructor(
    dataProvider: DataProvider,
    filesProvider: FilesProvider,
    resultSender: ResultSender,
    context: Context,
    testData?
  ) {
    this.dataProvider = dataProvider;
    this.filesProvider = filesProvider;
    this.resultSender = resultSender;
    this._context = context;
    this._logger = new Logger("ReportGenerator");
    this.testData = testData;
  }

  async generate(reportParamsId: any): Promise<IReportResult> {
    if (reportParamsId) {
      const tmpDir = dirSync();
      this._context.getRelativeFileFinder().setFolder(tmpDir.name);
      this._logger.debug(tmpDir.name);
    } else {
      this._context
        .getRelativeFileFinder()
        .setFolder(path.normalize(__dirname + `/../tests/unit/__mocks__/${this.testData?.templatesPath}`));
    }

    const data = await this.prepareData(reportParamsId);
    let items, reportParams, reportDefinition, itemFieldsStruct;
    if (data) {
      items = data.items;
      reportParams = data.reportParams;
      reportDefinition = data.reportDefinition;
      itemFieldsStruct = data.itemFieldsStruct;
    } else {
      return { valid: false, errorMessage: InputDataNotProvided };
    }
    /* Read the template metadata from JSON files */
    this._context.getTemplater().imgPath =
      path.resolve(this._context.getRelativeFileFinder().getFolder(), "efs-images/optimized") || "";
    /* Load JSON template data */
    StyleFields.loadData(this._context, reportDefinition["style"]);
    Placeholders.loadData(this._context, reportDefinition["placeholderMetadata"]);

    // Data
    if (reportParams["projectFields"]) {
      ProjectFields.loadData(this._context, reportParams["projectFields"]);
    }

    ItemsStruct.loadData(this._context, itemFieldsStruct);
    ReadyComponents.loadData(this._context, reportDefinition["readyComponents"]);
    Formulas.loadData(this._context, reportDefinition["formulas"]);

    Items.loadData(this._context, items);

    if(!reportDefinition["general"] || !reportDefinition["general"].templatePath) {
      return { valid: false, errorMessage: TemapltePathIsRequired };
    }
    /* Create the new document from the template */
    this._context
      .getTemplater()
      .create(
        reportDefinition["general"]?.templatePath,
        reportParams["attachments"] || [],
        reportDefinition["attachments"] || [],
        reportParams["conditionalPart"] || [],
        reportDefinition["settings"] || {}
      );

    /* Replace each markup for the real content */
    let docXmlText = "";
    try {
      docXmlText = this._context.getTemplater().replace();

    } catch (error) {
      const e = {
        message: error.message,
        name: error.name,
        stack: error.stack,
        properties: error.properties
      };
      this._logger.error("Error replazing markups", e);
      this.handleError(reportParamsId);
      return { valid: false };
    }

    /* Save the new document */
    const saveParams = this._context.getTemplater().getSaveParams();
    this.saveFile(reportParamsId, saveParams.fileName, saveParams.buffer);
    return { valid: true, reportId: reportParamsId, filename: saveParams.fileName, content: docXmlText };
  }

  handleError(reportParamsId: any) {
    if (reportParamsId) {
      this.resultSender.updateStatusFailed();
    }
  }

  saveFile(reportParamsId: any, fileName: any, buffer: any) {
    if (reportParamsId) {
      const s3 = new S3();
      const params = {
        Bucket: process.env.RESULTS_BUCKET,
        Key: fileName,
        Body: buffer
      };
      s3.upload(params, (err: any, data: any) => {
        if (err) {
          this._logger.error(err);
          throw err;
        }
        this.resultSender.updatePath(data.Location).then(() => {
          this._logger.debug("File uploaded successfully.", { location: data.Location });
        });
      });
    } else if(process.env.WRITE_LOCAL_TEST_FILE) {
        fs.writeFileSync(__dirname + "/../test/output/" + fileName, buffer);
    }
  }

  async prepareData(reportParamsId: any) {
    if (reportParamsId) {
      return await this.fetchData(reportParamsId);
    }

    return await this.testData;
  }

  private async fetchData(reportParamsId: any) {
    this.resultSender.updateStatusStarted();
    this._logger.debug("Starting to generate new report", { id: reportParamsId });
    const reportParams = await this.dataProvider.getReportParams(reportParamsId);
    if (
      !reportParams ||
      !reportParams.projectId ||
      !reportParams.accountId ||
      !reportParams.reportDefinitionId
    ) {
      this._logger.error("Report Data was not found or invalid", { error: reportParams });
      this.resultSender.updateStatusFailed();
      return undefined;
    }

    this._logger.debug("Getting data from remote server");
    const reportDefinition = await this.dataProvider.getReportDefinition(
      reportParams.reportDefinitionId
    );
    const items = await this.dataProvider.getItems(reportParams.projectId);
    const itemFieldsStruct = await this.dataProvider.getFields();
    this._logger.debug("Done getting data from remote server");

    /* get definition */
    if (!reportDefinition) {
      this._logger.error("Definition was not found");
      this.resultSender.updateStatusFailed();
      return undefined;
    }

    await new TemplateDownloader(this.filesProvider).download(
      reportDefinition,
      this._context.getRelativeFileFinder().getFolder()
    );
    return { items, reportParams, itemFieldsStruct, reportDefinition };
  }
}
