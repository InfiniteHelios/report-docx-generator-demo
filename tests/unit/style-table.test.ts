import { doesNotReject } from "assert";
import { IReportResult } from "../../src/report-generator";
import { Server } from "../../src/server";
import { cloneObject } from "./test-utils";
import testData from "./__mocks__/style-table/style-table.data";

describe("Table style", () => {
  it("Repeat header", async () => {
    const server = new Server();
    const localtestData = cloneObject(testData);
    localtestData.reportDefinition.style[0].style.repeatHeader = true;
    const result: IReportResult = await server.test(localtestData);
    expect(result.valid).toBeTruthy();
    expect(result.content).toMatchSnapshot();
  });
  
  it("Align right and fixed width", async () => {
    const server = new Server();
    const localtestData = cloneObject(testData);
    localtestData.reportDefinition.style[0].style.align = "right";
    localtestData.reportDefinition.style[0].style.fixedWidth = "300";
    localtestData.reportDefinition.readyComponents[0].fieldsToShow = [
      "NUMBER0",
      "TEXT1",
      "NUMBER1",
    ];
    const result: IReportResult = await server.test(localtestData);
    expect(result.valid).toBeTruthy();
    expect(result.content).toMatchSnapshot();
  });

  it("Align center and fixed width", async () => {
    const server = new Server();
    const localtestData = cloneObject(testData);
    localtestData.reportDefinition.style[0].style.align = "center";
    localtestData.reportDefinition.style[0].style.fixedWidth = "400";
    localtestData.reportDefinition.readyComponents[0].fieldsToShow = [
      "NUMBER0",
      "TEXT1",
      "NUMBER1",
    ];
    const result: IReportResult = await server.test(localtestData);
    expect(result.valid).toBeTruthy();
    expect(result.content).toMatchSnapshot();
  });

  it("Align left and fixed width", async () => {
    const server = new Server();
    const localtestData = cloneObject(testData);
    localtestData.reportDefinition.style[0].style.align = "left";
    localtestData.reportDefinition.style[0].style.fixedWidth = "200";
    localtestData.reportDefinition.readyComponents[0].fieldsToShow = [
      "NUMBER0",
      "TEXT1",
      "NUMBER1",
    ];
    const result: IReportResult = await server.test(localtestData);
    expect(result.valid).toBeTruthy();
    expect(result.content).toMatchSnapshot();
  });

  it("Global row height 50", async () => {
    const server = new Server();
    const localtestData = cloneObject(testData);
    localtestData.reportDefinition.style[0].style.rowHeight = "50";
    localtestData.reportDefinition.readyComponents[0].fieldsToShow = [
      "NUMBER0",
      "TEXT1",
      "NUMBER1",
    ];
    const result: IReportResult = await server.test(localtestData);
    expect(result.valid).toBeTruthy();
    expect(result.content).toMatchSnapshot();
  });

  it("Table border", async () => {
    const server = new Server();
    const localtestData = cloneObject(testData);
    localtestData.reportDefinition.style[0].style.tableBorder = [
      { type: "ALL", width: 4, color: "#e1e1e1" },
    ];
    const result: IReportResult = await server.test(localtestData);
    expect(result.valid).toBeTruthy();
    expect(result.content).toMatchSnapshot();
  });

  it("Cell border", async () => {
    const server = new Server();
    const localtestData = cloneObject(testData);
    localtestData.reportDefinition.style[0].style.cellBorder = [
      {
        width: 2,
        color: "yellow",
        condition: {
          field: "NUMBER2",
          type: "ALL",
        },
      },
      {
        width: 2,
        color: "red",
        condition: {
          field: "NUMBER2",
          type: "ABOVE",
          value: 99,
        },
      },
      {
        width: 2,
        color: "blue",
        condition: {
          field: "NUMBER2",
          type: "UNDER",
          value: 100,
        },
      },
      {
        width: 2,
        color: "pink",
        condition: {
          field: "NUMBER2",
          type: "EQUALS",
          value: 164,
        },
      },
      {
        width: 2,
        color: "pink",
        condition: {
          field: "NUMBER2",
          type: "EQUALS",
          value: 711,
        },
      },
    ];
    const result: IReportResult = await server.test(localtestData);
    expect(result.valid).toBeTruthy();
    expect(result.content).toMatchSnapshot();
  });
});
