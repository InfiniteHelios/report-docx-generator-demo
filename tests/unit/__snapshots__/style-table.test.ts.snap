"Sat Aug 22 2015 15:51:00 GMT-0700 (Pacific Daylight Time)" 
